include/core/physical_property_model.h

```diff
@@ -0,0 +1,174 @@
+ */
+class PhysicalPropertyModel
+{
+public:
+  /**
+   * @brief PhysicalPropertyModel Default constructor. Set the model_depends_on to false for all variable.
+   */
+  PhysicalPropertyModel()
+  {
+    model_depends_on[shear_rate]           = false;
+    model_depends_on[temperature]          = false;
+    model_depends_on[previous_temperature] = false;
+  }
+
+
+  /**
+   * @brief value Calculates the value of a physical property.
+   * @param fields_value Value of the various field on which the property may depend.
+   * @return value of the physical property calculated with the fields_value
+   */
+  virtual double
+  value(const std::map<field, double> &fields_value) = 0;
+
+  /**
+   * @brief vector_value Calculates the values of a physical property for
+   * @param field_vectors
+   */
+  virtual void
+  vector_value(const std::map<field, std::vector<double>> &field_vectors,
+               std::vector<double> &                       property_vector) = 0;
+
+  /**
+   * @brief jacobian Calcualtes the jacobian (the partial derivative) of the physical
+   * property with respect to a field
+   * @param field_values Value of the various fields on which the property may depend.
+   * @param id Indicator of the field with respect to which the jacobian
+   * should be calculated
+   * @return value of the partial derivative of the property with respect to the field.
+   */
+
+  virtual double
+  jacobian(const std::map<field, double> &field_values, const field id) = 0;
+
+  /**
+   * @brief vector_jacobian Calculate the derivative of the property with respect to a field
+   * @param field_vectors Vector for the values of the fields used to evaluated the property
+   * @param id Identifier of the field with respect to which a derivative should be calculated
+   * @param jacobian Vector of the value of the derivative of the property with respect to the field id
+   */
+
+  virtual void
+  vector_jacobian(const std::map<field, std::vector<double>> &field_vectors,
+                  const field                                 id,
+                  std::vector<double> &jacobian_vector) = 0;
+
+
+  /**
+   * @brief numerical_jacobian Calculate the jacobian through a finite difference approximation.
+   * This approach, although not preferable, is meant as a fall-back when
+   * calculating the jacobian manually is too difficult.
+   * @param fields_values The values of the various fields
+   * @param id Field id of the field with respect to which the jacobian should be calculated
+   * @return
+   */
+  inline double
+  numerical_jacobian(const std::map<field, double> &field_values,
+                     const field                    id)
+  {
+    double f_x   = this->value(field_values);
+    auto   x_dx  = field_values;
+    double dx    = std::max(1e-6 * x_dx[id], 1e-8);
+    x_dx[id]     = x_dx[id] + dx;
+    double f_xdx = this->value(x_dx);
+    return (f_xdx - f_x) / dx;
```

Comment:
Since you are using a linear approximation here, I think it's better to use the name "Euler" instead of "finite difference" in the description.

Reply:
Ok, I'll add both term to be clear.